staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go

```diff
@@ -241,26 +242,36 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 			}
 			metrics.WatchEvents.WithContext(req.Context()).WithLabelValues(kind.Group, kind.Version, kind.Kind).Inc()
 
-			obj := s.Fixup(event.Object)
-			if err := embeddedEncodeFn(obj, buf); err != nil {
-				// unexpected error
-				utilruntime.HandleError(fmt.Errorf("unable to encode watch object %T: %v", obj, err))
-				return
-			}
+			// Returns an external representation of a watch event that will be then serialized
+			// and sent to the watch client.
+			newOutEvent = func() (runtime.Object, error) {
+				obj := s.Fixup(event.Object)
+				if err := embeddedEncodeFn(obj, buf); err != nil {
+					// unexpected error
+					return nil, fmt.Errorf("unable to encode watch object %T: %v", obj, err)
+				}
 
-			// ContentType is not required here because we are defaulting to the serializer
-			// type
-			unknown.Raw = buf.Bytes()
-			event.Object = &unknown
-			metrics.WatchEventsSizes.WithContext(req.Context()).WithLabelValues(kind.Group, kind.Version, kind.Kind).Observe(float64(len(unknown.Raw)))
+				// ContentType is not required here because we are defaulting to the serializer
+				// type
+				unknown.Raw = buf.Bytes()
+				event.Object = &unknown
 
-			*outEvent = metav1.WatchEvent{}
+				metrics.WatchEventsSizes.WithContext(req.Context()).WithLabelValues(kind.Group, kind.Version, kind.Kind).Observe(float64(len(unknown.Raw)))
```

Comment:
This must be exported for each watcher separately, not just once.

What we can do instead is:
* we can change streaming.NewEncoder to count bytes after marshalling and expose metric, or
* write another io.Writer between streaming.NewEncoder and framer that implements Write(buf []byte) that passes it to enother io.Writer, but counts len(buf) first and exposes metric ("BytesCountingWriter").

Reply:
This will slightly change meaning of the metric (it will include metadata of watchevents objects, not only len of Raw object inside, but maybe it's actually even better this way